dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.cpp

```diff
@@ -0,0 +1,414 @@
+        scheduler_need_wakeup = true;
+        scheduler_notifier.notify_all();
+    }
+
+    if (is_started)
+        scheduler_thread.join();
+
+    // Then wait all running tasks to finish.
+    pool.reset();
+
+    LOG_INFO(logger, "LocalIndexerScheduler is destroyed");
+}
+
+void LocalIndexerScheduler::start()
+{
+    if (is_started)
+        return;
+
+    scheduler_thread = std::thread([this]() { schedulerLoop(); });
+    is_started = true;
+}
+
+void LocalIndexerScheduler::waitForFinish()
+{
+    while (true)
+    {
+        std::unique_lock lock(mutex);
+        if (all_tasks_count == 0 && running_tasks_count == 0)
+            return;
+        on_finish_notifier.wait(lock);
+    }
+}
+
+void LocalIndexerScheduler::pushTask(const Task & task)
+{
+    if (pool_max_memory_limit > 0 && task.request_memory > pool_max_memory_limit)
+        throw Exception(fmt::format(
+            "Requests memory exceeds limit (request={} limit={})",
+            task.request_memory,
+            pool_max_memory_limit));
+
+    std::unique_lock lock(mutex);
+
+    const auto internal_task = std::make_shared<InternalTask>(InternalTask{
+        .user_task = task,
+        .created_at = Stopwatch(),
+        .scheduled_at = Stopwatch(), // Not scheduled
+    });
+
+    // Whether task is ready is undertermined. It can be changed any time
+    // according to current running tasks.
+    // The scheduler will find a better place for this task when meeting it.
+    ready_tasks[task.keyspace_id][task.table_id].emplace_back(internal_task);
+    ++all_tasks_count;
+
+    scheduler_need_wakeup = true;
+    scheduler_notifier.notify_all();
+}
+
+size_t LocalIndexerScheduler::dropTasks(KeyspaceID keyspace_id, TableID table_id)
+{
+    size_t dropped_tasks = 0;
+
+    std::unique_lock lock(mutex);
+    if (auto it = ready_tasks.find(keyspace_id); it != ready_tasks.end())
+    {
+        auto & tasks_by_table = it->second;
+        if (auto table_it = tasks_by_table.find(table_id); table_it != tasks_by_table.end())
+        {
+            dropped_tasks += table_it->second.size();
+            tasks_by_table.erase(table_it);
+        }
+        if (tasks_by_table.empty())
+            ready_tasks.erase(it);
+    }
+    for (auto it = unready_tasks.begin(); it != unready_tasks.end();)
+    {
+        if ((*it)->user_task.keyspace_id == keyspace_id && (*it)->user_task.table_id == table_id)
+        {
+            it = unready_tasks.erase(it);
+            ++dropped_tasks;
+        }
+        else
+        {
+            it++;
+        }
+    }
+
+    LOG_INFO(logger, "Removed {} tasks, keyspace_id={} table_id={}", dropped_tasks, keyspace_id, table_id);
+
+    return dropped_tasks;
+}
+
+bool LocalIndexerScheduler::isTaskReady(std::unique_lock<std::mutex> &, const InternalTaskPtr & task)
+{
+    for (const auto & file_id : task->user_task.file_ids)
+    {
+        if (adding_index_page_id_set.find(file_id) != adding_index_page_id_set.end())
```

Comment:
Why don't make unready_tasks into a set, and check unready_tasks?

Reply:
Since it will be fair which means the task added firstly should be executed firstly.